python/llm/src/bigdl/llm/transformers/models/mpt.py

```diff
@@ -0,0 +1,139 @@
+                                                           softmax_scale=self.softmax_scale, attn_bias=attn_bias,
+                                                           key_padding_mask=key_padding_mask, is_causal=is_causal,
+                                                           dropout_p=self.attn_dropout_p, training=self.training,
+                                                           needs_weights=needs_weights)
+    return (self.out_proj(context), attn_weights, past_key_value)
+
+
+def mpt_scaled_multihead_dot_product_attention(query, key, value, n_heads,
+                                               past_key_value=None,
+                                               softmax_scale=None,
+                                               attn_bias=None,
+                                               key_padding_mask=None,
+                                               is_causal=False,
+                                               dropout_p=0.0,
+                                               training=False,
+                                               needs_weights=False,
+                                               multiquery=False):
+    q = rearrange(query, 'b s (h d) -> b h s d', h=n_heads)
+    bsz, n_heads, q_len, head_dim = q.size()
+    device = q.device
+    kv_n_heads = 1 if multiquery else n_heads
+    k = rearrange(key, 'b s (h d) -> b h d s', h=kv_n_heads)
+    v = rearrange(value, 'b s (h d) -> b h s d', h=kv_n_heads)
+    kv_seq_len = k.shape[-1]
+    if past_key_value is not None:
+        if len(past_key_value) != 0:
+        #     k = torch.cat([past_key_value[0], k], dim=3)
+        #     v = torch.cat([past_key_value[1], v], dim=2)
+            cache_k = past_key_value[0].permute(0, 1, 3, 2)
+            cache_v = past_key_value[1]
+            kv_seq_len += cache_k.shape[-2]
+            if cache_k.stride()[1] <= cache_k.size(2) * cache_k.size(3):
+                if device.type == 'xpu':
+                    torch.xpu.empty_cache()
+                # allocate new
+                new_cache_k, new_cache_v = create_kv_cache(bsz,
+                                                           kv_n_heads,  # Support GQA
+                                                           head_dim,
+                                                           cache_k.size(2),
+                                                           kv_seq_len + KV_CACHE_ALLOC_BLOCK_LENGTH,
+                                                           dtype=cache_k.dtype,
+                                                           device=device)
+                new_cache_k[:] = cache_k
+                new_cache_v[:] = cache_v
+                cache_k = new_cache_k
+                cache_v = new_cache_v
+            key_states, value_states = append_kv_cache(cache_k, cache_v, k.permute(0, 1, 3, 2), v)
+            k = key_states.permute(0, 1, 3, 2)
+            v = value_states
+        else:
```

Comment:
maybe `elif use_cache:`?

Reply:
There is no use_cache flag in this model. In this model, just check whether len(past_key_value) != 0 to see if it is the first time since initially past_key_value=()